synapse/storage/persist_events.py

```diff
@@ -734,7 +746,69 @@ async def _get_new_state_after_events(
             state_res_store=StateResolutionStore(self.main_store),
         )
 
-        return res.state, None
+        # If the returned state matches the state group of one of the new
+        # forward extremities then we check if we are able to prune some state
+        # extremities.
+        if res.state_group and res.state_group in new_state_groups:
+            # We keep all the extremities that have the same state group, and
+            # see if we can drop the others.
+            new_new_extrems = {
+                e
+                for e in new_latest_event_ids
+                if event_id_to_state_group[e] == res.state_group
+            }
+
+            dropped = set(new_latest_event_ids) - new_new_extrems
+
+            # We only drop events if:
+            #   1. we're not currently persisting them;
+            #   2. they're not our own events (or are dummy events); and
+            #   3. they're either over N minutes old or we have newer events
+            #      from the same domain.
+            #
+            # The idea is that we don't want to drop events that are "legitmate"
+            # extremities (that we would want to include as prev events), only
+            # "stuck" extremities that are e.g. due to a gap in the graph.
+            #
+            # Note that we either drop all of them or none of them. If we only
+            # drop some of the events we don't know if state res would come to the same conclusion.
+
+            drop = True
+            for ev, _ in events_context:
+                if ev.event_id in dropped:
+                    drop = False
+                    break
+
+            if drop:
+                dropped_events = await self.main_store.get_events(
+                    dropped,
+                    allow_rejected=True,
+                    redact_behaviour=EventRedactBehaviour.AS_IS,
+                )
+
+                new_senders = {e.sender for e, _ in events_context}
+
+                twenty_minutes_ago = self._clock.time_msec() - 20 * 60 * 1000
+                for event in dropped_events.values():
+                    if (
+                        self.is_mine_id(event.sender)
+                        and event.type != "org.matrix.dummy_event"
+                    ):
+                        drop = False
+                        break
+
+                    if event.origin_server_ts < twenty_minutes_ago:
```

Comment:
I'm not really sure if we should be dropping based on age, and twenty minutes may be way too aggressive.

Reply:
Ok, I've made this way less aggressive.